File: apps/spell-circle-canvas/src/common/compose/reference/examples/background_verb.py
"""background — a decoration painted beneath the fill, which is where a
shadow and anything else the surface sits on top of belongs.

The Python twin of background_verb.cpp. A reference example: it is
rendered with ``sigil render`` and belongs to no sketch registry.
"""

from sigil import compose
from sigil.sketch import SketchContext, sketch

GROUND = "#1a2027"
PLATE = "#e9eef1"
SHADE = "#05080b8c"
ASH = "#8ea0ad"


def cell(caption: str, plate: compose.Element) -> compose.Element:
    return (
        compose.box(
            plate.width(160).height(100).borderRadius(12),
            compose.text(caption, size=12, color=ASH),
        )
        .column()
        .gap(12)
        .basis(0)
        .grow(1)
        .alignItems("center")
        .justifyContent("center")
    )


@sketch(size=(620, 250), background=GROUND, capture_at=0)
class BackgroundVerb:
    def setup(self, ctx: SketchContext) -> None:
        ctx.render(self.describe())

    def describe(self) -> compose.Element:
        return (
            compose.box(
                cell("the fill alone", compose.box().fill(PLATE)),
                # Beneath the fill: the plate sits on the shadow.
                cell(
                    "background(shadow)",
                    compose.box()
                    .fill(PLATE)
                    .background(compose.shadow(SHADE, (0, 10), 18)),
                ),
                # The same mark with nothing over it — a background under
                # no fill is the whole of what the node paints.
                cell(
                    "background with no fill",
                    compose.box().background(compose.shadow(SHADE, (0, 10), 18)),
                ),
            )
            .row()
            .gap(18)
            .padding(28)
        )

File: apps/spell-circle-canvas/src/common/compose/reference/examples/background_verb.cpp
/** @file
 * background — a decoration painted beneath the fill, which is where a
 * shadow and anything else the surface sits on top of belongs.
 *
 * A reference example. It stands outside the sketch registry, so it is
 * photographed with `--frame` and never enters the plate sweep.
 */

#include <sigilcompose/brush/Decorations.h>
#include <sigilcompose/core/Core.h>
#include <sigilmaterial/color/Color.h>
#include <sigilsketch/canvas/Sketch.h>

namespace material = sigil::material;
namespace sketch = sigil::sketch;

using namespace sigil::compose;

namespace {

constexpr SkSize kCanvas = {620, 250};
constexpr material::Color kGround = hexColor(0x1a2027);
constexpr material::Color kPlate = hexColor(0xe9eef1);
constexpr material::Color kShade = hexColor(0x05080b, 0.55f);
constexpr material::Color kAsh = hexColor(0x8ea0ad);

Element cell(const char* caption, Element plate) {
  return box()
      .column()
      .gap(12)
      .flexBasis(0)
      .flexGrow(1)
      .alignItems(Align::Center)
      .justifyContent(Justify::Center)
      .children({plate.width(160).height(100).borderRadius({12}),
                 text(caption).font(
                     {.size = 12, .color = material::skia::toSkColor(kAsh)})});
}

}  // namespace

struct BackgroundVerb {
  void setup(sketch::SketchContext& ctx) {
    ctx.canvas({.size = kCanvas, .background = kGround, .captureSeconds = 0});
    ctx.composer.render(describe());
  }

  Element describe() const {
    return box().row().gap(18).padding(28).children({
        cell("the fill alone", box().fill(kPlate)),
        // Beneath the fill: the plate sits on the shadow.
        cell("background(shadow)",
             box().fill(kPlate).background(shadow(kShade, {0, 10}, 18))),
        // The same mark with nothing over it — a background under no fill
        // is the whole of what the node paints.
        cell("background with no fill",
             box().background(shadow(kShade, {0, 10}, 18))),
    });
  }
};

SIGIL_SKETCH(BackgroundVerb, "Reference · Compose",
             "a shadow beneath the fill, and the same mark with no fill "
             "over it")
